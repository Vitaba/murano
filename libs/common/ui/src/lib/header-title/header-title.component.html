<ng-container *ngIf="!errors; else showErrors">
<div
  [class]="styles.container">
  <ng-template #titleDefaultTemplate>
    <h1 [class]="styles.title">{{data.title}}</h1>
  </ng-template>

  <ng-container [ngTemplateOutlet]="titleExtraTemplate || titleDefaultTemplate">
  </ng-container>

  <ng-template #descriptionDefaultTemplate>
    <div [class]="styles.description">{{data.description}}</div>
  </ng-template>

  <ng-container [ngTemplateOutlet]="descriptionExtraTemplate || descriptionDefaultTemplate">
  </ng-container>

  <ng-template #lineDefaultTemplate>
    <hr [class]="styles.line" />
  </ng-template>

  <ng-container [ngTemplateOutlet]="lineExtraTemplate || lineDefaultTemplate">
  </ng-container>

</div>
</ng-container>

<ng-template #showErrors>
  {{ errors | json}}
</ng-template>
