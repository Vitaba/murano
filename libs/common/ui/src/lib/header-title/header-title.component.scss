@import 'libs/scss/ui/src/lib/base/base.scss';
