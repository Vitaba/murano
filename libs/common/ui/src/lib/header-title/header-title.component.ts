import { AfterViewInit, ChangeDetectionStrategy, ChangeDetectorRef, Component, ContentChild, Input, OnChanges, SimpleChanges, TemplateRef } from '@angular/core';
import { AvatarData, GuardError, HeaderData, HeaderStyles, isHeaderData } from '@vitaba/common-utils';
import { BehaviorSubject } from 'rxjs';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'vitaba-header-title',
  styleUrls: ['./header-title.component.scss'],
  templateUrl: './header-title.component.html',
})
export class HeaderTitleComponent implements OnChanges, AfterViewInit {
  public changes!: BehaviorSubject<SimpleChanges>;
  public errors!: GuardError;

  @Input() public styles: HeaderStyles = {
    container: `font-sans font-light leading-normal markdown mb-6 px-6
    max-w-3xl mx-auto lg:ml-0 lg:mr-auto xl:mx-0 xl:px-12 xl:w-3/4`,
    description: 'mt-0 mb-4 text-gray-600',
    line: 'my-8 border-b-2 border-gray-200',
    title: 'text-3xl text-gray-900',
  };
  @Input() public data: HeaderData = {
    description: 'Useful resources and assets for users.',
    title: 'News',
  };

  @ContentChild('titleExtraTemplate', { static: false })
  // tslint:disable-next-line:no-any
  public titleExtraTemplate!: TemplateRef<any>;

  @ContentChild('descriptionExtraTemplate', { static: false })
  // tslint:disable-next-line:no-any
  public descriptionExtraTemplate!: TemplateRef<any>;

  @ContentChild('lineExtraTemplate', { static: false })
  // tslint:disable-next-line:no-any
  public lineExtraTemplate!: TemplateRef<any>;

  public constructor(
    private readonly _changeDetectorRef: ChangeDetectorRef,
  ) {}

  public ngOnChanges(changes: SimpleChanges): void {
    if (!this.changes) {
      this.changes = new BehaviorSubject(changes);
    }
    this.changes.next(changes);
  }

  public ngAfterViewInit(): void {
    this.changes.subscribe((changes: SimpleChanges) => {
      if (changes.data) {
        const validation =
        this.validateHeaderData(changes.data.currentValue as HeaderData);

        if (validation) {
          this.errors = { ...this.errors, ...validation };
        }
      }
      if (changes.styles) {
        const validation =
        this.validateHeaderStyles(changes.styles.currentValue as HeaderStyles);

        if (validation) {
          this.errors = { ...this.errors, ...validation };
        }
      }

      this._changeDetectorRef.detectChanges();
    });
  }

  public validateHeaderData(value: HeaderData) {
    const validations = [];

    if (!this.titleExtraTemplate) {
      validations.push({
        name: 'title',
        type: 'string',
      });
    }

    if (!this.descriptionExtraTemplate) {
      validations.push({
        name: 'description',
        type: 'string',
      });
    }

    const validation = isHeaderData(value, validations);

    return !validation.valid ? validation : undefined;
  }

  public validateHeaderStyles(value: HeaderStyles) {
    const validations = [];

    if (!this.titleExtraTemplate) {
      validations.push({
        name: 'title',
        type: 'string',
      });
    }

    if (!this.descriptionExtraTemplate) {
      validations.push({
        name: 'description',
        type: 'string',
      });
    }
    if (!this.lineExtraTemplate) {
      validations.push({
        name: 'line',
        type: 'string',
      });
    }

    const validation = isHeaderData(value, validations);

    return !validation.valid ? validation : undefined;
  }
}
