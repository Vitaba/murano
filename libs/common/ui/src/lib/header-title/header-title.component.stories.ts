// tslint:disable:no-default-export object-literal-sort-keys ter-max-len

import { text } from '@storybook/addon-knobs';
import { HeaderTitleComponent } from './header-title.component';
export default {
  title: 'Header Title ',
};

const data = {
  description: 'Useful resources and assets for users.',
  title: 'News',
};

const styles = {
  container: 'font-sans font-light leading-normal markdown mb-6 px-6 max-w-3xl mx-auto lg:ml-0 lg:mr-auto xl:mx-0 xl:px-12 xl:w-3/4',
  description: 'mt-0 mb-4 text-gray-600',
  line: 'my-8 border-b-2 border-gray-200',
  title: 'text-3xl text-gray-900',
};

export const basic = () => ({
  moduleMetadata: {
    declarations: [HeaderTitleComponent],
    imports: [],
  },
  props: {
    data_title: text('title', data.title, 'data'),
    data_description: text('description', data.description, 'data'),
    styles_container: text('container', styles.container, 'styles'),
    styles_title: text('title', styles.title, 'styles'),
    styles_description: text('description', styles.description, 'styles'),
    styles_line: text('line', styles.line, 'styles'),
  },
  template: `
  <vitaba-header-title
    [data]="{
    title: data_title,
    description: data_description
    }"
    [styles]="{
    container: styles_container,
    description: styles_description,
    line: styles_line,
    title: styles_title
    }"
    > </vitaba-header-title>
  `,
});
