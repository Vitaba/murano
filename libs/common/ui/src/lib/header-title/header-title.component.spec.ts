
// tslint:disable: no-any component-max-inline-declarations ter-max-len
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { HeaderTitleComponent } from './header-title.component';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'vitaba-header-title-sample-implementation',
  template: `
  <ng-container [ngSwitch]="mode">
    <vitaba-header-title *ngSwitchDefault [data]="data" [styles]="styles">
    </vitaba-header-title>

    <vitaba-header-title *ngSwitchCase="'title'" [data]="data" [styles]="styles">
      <ng-template #titleExtraTemplate>
        <p>Title content-projection</p>
      </ng-template>
    </vitaba-header-title>

    <vitaba-header-title *ngSwitchCase="'description'" [data]="data" [styles]="styles">
      <ng-template #descriptionExtraTemplate>
        <p>Description content-projection</p>
      </ng-template>
    </vitaba-header-title>

    <vitaba-header-title *ngSwitchCase="'line'" [data]="data" [styles]="styles">
      <ng-template #lineExtraTemplate>
        <p>Line content-projection</p>
      </ng-template>
    </vitaba-header-title>
  </ng-container>
  `,
})
class TestHostComponent {
  public styles = {
    container: `font-sans font-light leading-normal markdown mb-6 px-6
    max-w-3xl mx-auto lg:ml-0 lg:mr-auto xl:mx-0 xl:px-12 xl:w-3/4`,
    description: 'mt-0 mb-4 text-gray-600',
    line: 'my-8 border-b-2 border-gray-200',
    title: 'text-3xl text-gray-900',
  };
  public data = {
    description: 'Useful resources and assets for users.',
    title: 'Lorem Ipsum',
  };
  public mode = 'default';
}

describe('HeaderTitleComponent', () => {
  let component: TestHostComponent;
  let fixture: ComponentFixture<TestHostComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TestHostComponent, HeaderTitleComponent],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestHostComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  it('should be default', () => {
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should be titleExtraTemplate', () => {
    component.mode = 'title';
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should be descriptionExtraTemplate', () => {
    component.mode = 'description';
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should be lineExtraTemplate', () => {
    component.mode = 'line';
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should throw errors when data is not in right format', () => {
    component.mode = 'default';
    component.data = {
      description: 1,
      title: 1,
    } as any;
    fixture.changeDetectorRef.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
