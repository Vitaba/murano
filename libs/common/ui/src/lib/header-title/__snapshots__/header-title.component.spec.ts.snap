// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderTitleComponent should be default 1`] = `
<vitaba-header-title-sample-implementation
  data={[Function Object]}
  mode={[Function String]}
  styles={[Function Object]}
>
  <vitaba-header-title
    ng-reflect-data="[object Object]"
    ng-reflect-styles="[object Object]"
  >
    
    
    <div
      class="font-sans font-light leading-normal markdown mb-6 px-6
      max-w-3xl mx-auto lg:ml-0 lg:mr-auto xl:mx-0 xl:px-12 xl:w-3/4"
    >
      
      
      <h1
        class="text-3xl text-gray-900"
      >
        Lorem Ipsum
      </h1>
      
      
      <div
        class="mt-0 mb-4 text-gray-600"
      >
        Useful resources and assets for users.
      </div>
      
      
      <hr
        class="my-8 border-b-2 border-gray-200"
      />
    </div>
    
  </vitaba-header-title>
</vitaba-header-title-sample-implementation>
`;

exports[`HeaderTitleComponent should be descriptionExtraTemplate 1`] = `
<vitaba-header-title-sample-implementation
  data={[Function Object]}
  mode={[Function String]}
  styles={[Function Object]}
>
  <vitaba-header-title
    ng-reflect-data="[object Object]"
    ng-reflect-styles="[object Object]"
  >
    
    
    <div
      class="font-sans font-light leading-normal markdown mb-6 px-6
      max-w-3xl mx-auto lg:ml-0 lg:mr-auto xl:mx-0 xl:px-12 xl:w-3/4"
    >
      
      
      <h1
        class="text-3xl text-gray-900"
      >
        Lorem Ipsum
      </h1>
      
      
      <p>
        Description content-projection
      </p>
      
      
      <hr
        class="my-8 border-b-2 border-gray-200"
      />
    </div>
    
  </vitaba-header-title>
</vitaba-header-title-sample-implementation>
`;

exports[`HeaderTitleComponent should be lineExtraTemplate 1`] = `
<vitaba-header-title-sample-implementation
  data={[Function Object]}
  mode={[Function String]}
  styles={[Function Object]}
>
  <vitaba-header-title
    ng-reflect-data="[object Object]"
    ng-reflect-styles="[object Object]"
  >
    
    
    <div
      class="font-sans font-light leading-normal markdown mb-6 px-6
      max-w-3xl mx-auto lg:ml-0 lg:mr-auto xl:mx-0 xl:px-12 xl:w-3/4"
    >
      
      
      <h1
        class="text-3xl text-gray-900"
      >
        Lorem Ipsum
      </h1>
      
      
      <div
        class="mt-0 mb-4 text-gray-600"
      >
        Useful resources and assets for users.
      </div>
      
      
      <p>
        Line content-projection
      </p>
    </div>
    
  </vitaba-header-title>
</vitaba-header-title-sample-implementation>
`;

exports[`HeaderTitleComponent should be titleExtraTemplate 1`] = `
<vitaba-header-title-sample-implementation
  data={[Function Object]}
  mode={[Function String]}
  styles={[Function Object]}
>
  <vitaba-header-title
    ng-reflect-data="[object Object]"
    ng-reflect-styles="[object Object]"
  >
    
    
    <div
      class="font-sans font-light leading-normal markdown mb-6 px-6
      max-w-3xl mx-auto lg:ml-0 lg:mr-auto xl:mx-0 xl:px-12 xl:w-3/4"
    >
      
      
      <p>
        Title content-projection
      </p>
      
      
      <div
        class="mt-0 mb-4 text-gray-600"
      >
        Useful resources and assets for users.
      </div>
      
      
      <hr
        class="my-8 border-b-2 border-gray-200"
      />
    </div>
    
  </vitaba-header-title>
</vitaba-header-title-sample-implementation>
`;

exports[`HeaderTitleComponent should throw errors when data is not in right format 1`] = `
<vitaba-header-title-sample-implementation
  data={[Function Object]}
  mode={[Function String]}
  styles={[Function Object]}
>
  <vitaba-header-title
    ng-reflect-data="[object Object]"
    ng-reflect-styles="[object Object]"
  >
    
     {
    "valid": false,
    "guard": "isHeaderData",
    "value": {
      "description": 1,
      "title": 1
    },
    "message": "Is not a header data",
    "errors": [
      {
        "valid": false,
        "guard": "hasProperties",
        "value": {
          "description": 1,
          "title": 1
        },
        "errors": [
          {
            "value": {
              "name": "title",
              "type": "string"
            },
            "guard": "String Validation",
            "message": "Is not a string",
            "valid": false
          },
          {
            "value": {
              "name": "description",
              "type": "string"
            },
            "guard": "String Validation",
            "message": "Is not a string",
            "valid": false
          }
        ]
      }
    ]
  }
  
    
  </vitaba-header-title>
</vitaba-header-title-sample-implementation>
`;
